lib/encoding/yaml/encoder.toit

```diff
@@ -0,0 +1,273 @@
+
+  abstract encode-string_ str
+  abstract encode-number_ number
+  abstract encode-true_
+  abstract encode-false_
+  abstract encode-null_
+  abstract encode-map_ map [converter]
+  abstract encode-list_ list [converter]
+
+  /**
+  Outputs a list-like thing to the serialized stream.
+  This can be used by converter blocks.
+  The generator is called repeatedly with indices from 0 to size - 1.
+  */
+  abstract put-list size/int [generator] [converter]
+
+  put-unicode-escape_ code-point/int:
+    put-byte_ 'u'
+    put-byte_
+      to-lower-case-hex (code-point >> 12) & 0xf
+    put-byte_
+      to-lower-case-hex (code-point >> 8) & 0xf
+    put-byte_
+      to-lower-case-hex (code-point >> 4) & 0xf
+    put-byte_
+      to-lower-case-hex code-point & 0xf
+
+
+class Buffer_:
+  buffer_ := ByteArray INITIAL-BUFFER-SIZE_
+  offset_ := 0
+
+  to-string:
+    return buffer_.to-string 0 offset_
+
+  to-byte-array:
+    return buffer_.copy 0 offset_
+
+  ensure_ size -> none:
+    if offset_ + size <= buffer_.size: return
+    new-size := buffer_.size * 2
+    while new-size < offset_ + size:
+      new-size *= 2
+    new := ByteArray new-size
+    new.replace 0 buffer_
+    buffer_ = new
+
+  /**
+  Outputs a string or ByteArray directly to the JSON stream.
+  No quoting, no escaping.  This is mainly used for things
+    that will be parsed as numbers or strings by the receiver.
+  */
+  put-unquoted data -> none:
+    len := data.size
+    ensure_ len
+    buffer_.replace offset_ data
+    offset_ += len
+
+  put-string_ str from to:
+    len := to - from
+    ensure_ len
+    buffer_.replace offset_ str from to
+    offset_ += len
+
+  put-byte_ byte:
+    ensure_ 1
+    buffer_[offset_++] = byte
+
+  peek-last-byte_:
+    if offset_ > 0: return buffer_[offset_ - 1]
+    return null
+
+  clear_:
+    offset_ = 0
+
+class YamlEncoder extends EncoderBase_:
+  current-line-start-offset_/int := 0
+  indent_/int := 0
+  enclosed_in_map_ := false
+  indent_buffer_/string := "      " // A buffer of spaces, extends as nescessary
+
+  /** See $EncoderBase_.encode */
+  encode obj/any converter/Lambda:
+    return super obj converter
+
+  /** See $Buffer_.put-unquoted */
+  put-unquoted data -> none:
+    super data
+
+  put-value_ val/string:
+    if enclosed_in_map_: put-byte_ ' '
+    put-unquoted val
+
+  encode-string_ str/string:
+    // To determine if the str needs to be double quoted, we try to parse it, and if that works
+    // then it is safe to not double qoute it. Also double quotes all multi-line strings
+    should_quote := (catch: if not (parse str) is string: throw "ERROR") or
+                    str.contains "\n" or
+                    str.contains "\r"
+
+    escaped := escape-string str
+
+    if enclosed_in_map_: put-byte_ ' '
+    if should_quote: put-byte_ '"'
+    put-unquoted (escape-string str)
+    if should_quote: put-byte_ '"'
+
+  encode-number_ number:
+    str := number is float ? number.stringify 2 : number.stringify
```

Comment:
This is also used for the json encoder and changing to arbitrary precession will be a potentail breaking change, which is not the intention of this PR.

Reply:
Not asking to change. Just add a comment what YAML wants here.
Our double-conversion isn't correct anyway, but it would be nice to know what should happen here.